core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -222,4 +230,78 @@ public ByteSizeValue paramAsSize(String key, ByteSizeValue defaultValue) {
+     * {@code null}. Only use this if you need to do something different if the request is plain text. Use
+     * {@link #contentOrSourceParamParser()} if you need a parser.
+     */
+    @Nullable
+    public final XContentType contentOrSourceParamXContentType() {
+        BytesReference content = contentOrSourceParam();
+        return content == null ? null : XContentFactory.xContentType(content);
+    }
+
+    /**
+     * A parser for the contents of this request if it has contents, otherwise a parser for the {@code source} parameter if there is one,
+     * otherwise throws an {@link ElasticsearchParseException}. Use {@link #withContentOrSourceParamParserOrNull(IOConsumer)} instead if you
+     * need to handle the absence request content gracefully.
+     */
+    public final XContentParser contentOrSourceParamParser() throws IOException {
+        BytesReference content = contentOrSourceParam();
+        if (content.length() == 0) {
+            throw new ElasticsearchParseException("Body required");
+        }
+        return XContentFactory.xContent(content).createParser(content);
+    }
+
+    /**
+     * Call a consumer with the parser for the contents of this request if it has contents, otherwise with a parser for the {@code source}
+     * parameter if there is one, otherwise with {@code null}. Use {@link #contentOrSourceParamParser()} if you should throw an exception
+     * back to the user when there isn't request content.
+     */
+    public final void withContentOrSourceParamParserOrNull(IOConsumer<XContentParser> withParser) throws IOException {
+        XContentParser parser = null;
+        BytesReference content = contentOrSourceParam();
+        if (content.length() > 0) {
+            parser = XContentFactory.xContent(content).createParser(content);
+        }
+
+        try {
+            withParser.accept(parser);
+        } finally {
+            IOUtils.close(parser);
+        }
+    }
+
+    /**
+     * Get the content of the request or the contents of the {@code source} param. Prefer {@link #contentOrSourceParamParser()} if possible
+     * prefer methods like {@link #contentOrSourceParamParser()} and {@link #contentOrSourceParamXContentType()}. This should be private
+     * but isn't to support mostly legacy APIs.
+     */
+    public final BytesReference contentOrSourceParam() {
+        if (hasContent()) {
+            return content();
+        }
+        String source = param("source");
+        if (source != null) {
+            return new BytesArray(source);
+        }
+        return BytesArray.EMPTY;
+    }
+
+    /**
+     * The String representation of the body if there is one or the {@code source} parameter if there isn't content. Use this only if you
+     * aren't need to process raw strings.
+     */
+    public final String contentOrSourceParamString() throws IOException {
```

Comment:
I added it when I thought I could get away with making `contentOrSourceParam` private or at least deprecated. Now that I've pretty much given up on that we don't need this. In fact I can probably drop `contentOrSourceParamXContentType` as well under the same logic. Good intentions led to bad results I think.